assayist/processor/container_go_analyzer.py

```diff
@@ -0,0 +1,202 @@
+            return
+
+        # First process the source code that's directly available in
+        # the dist-git repository.
+        dist_git_excludes = [
+            # Files expected to be local and not upstream.
+            '.git*',
+            'Makefile',
+            'sources',
+            'Dockerfile*',
+            'additional-tags',
+            'content_sets.yml',
+            'container.yaml',
+            '.oit',
+            'public',
+            'scripts',
+        ]
+        srcdir = os.path.join(self.input_dir, self.SOURCE_DIR)
+        expected_exes = self._process_source_code(source_location, srcdir,
+                                                  excludes=dist_git_excludes)
+
+        # Next process source code from archives (from 'rhpkg sources').
+        # Look for tar archives and zip archives.
+        tar_archives = glob(os.path.join(srcdir, '*.tar.*'))
+        zip_archives = glob(os.path.join(srcdir, '*.zip'))
+        archives = [(unpack_tar, archive) for archive in tar_archives]
+        archives += [(unpack_zip, archive) for archive in zip_archives]
+        for unpack, archive in archives:
+            with tempfile.TemporaryDirectory() as subsrc:
+                unpack(archive, subsrc)
+                expected_exes += self._process_source_code(source_location, subsrc)
+
+        # Now claim all the Go executables with names matching the
+        # source code.
+        self._claim_go_executables(expected_exes)
+
+    def _process_source_code(self, source_location, srcdir, excludes=None):
+        """Run backvendor on the source code.
+
+        :param SourceLocation source_location: local source code DB node
+        :param str path: path to source code to examine
+        :param list excludes: list of globs to ignore
+        :return: set of expected executable basenames
+        :rtype: set
+        """
+        # Run backvendor.
+        with tempfile.NamedTemporaryFile() as excludes_file:
+            options = ['-template', self.BACKVENDOR_TEMPLATE]
+            if excludes:
+                excludes_file.write(''.join('%s\n' % e for e in excludes))
+                excludes_file.flush()
+                options += ['-excludes', excludes_file]
+
+            cmd = [self.BACKVENDOR] + options + ['.']
+            log.info(f'Running "{" ".join(cmd)}"')
+            bv = subprocess.Popen(cmd, cwd=srcdir,
+                                  stdout=subprocess.PIPE,
+                                  stderr=subprocess.PIPE)
+
+            (stdout, stderr) = bv.communicate()
+
+        returncode = bv.wait()
+        if returncode:
+            raise RuntimeError(f'The command "{" ".join(cmd)}" failed with: {stderr}')
+
+        expected_exes = set()
+
+        # There may be many top-level repositories, but we are only
+        # allowed a single 'upstream' relationship type. Only report
+        # the first top-level repository as "the upstream" one.
```

Comment:
Does it makes sense to create source locations for all the top-level repositories and then create embedded relationships to the internal source location instead?

Reply:
Instead of marking any of them as upstream, you mean?

Reply:
Yes

Reply:
Are you suggesting doing that in the case that there are multiple top-level repositories only, or also when there is only one top-level repository?

Reply:
If I'm understanding the situation correctly, we're talking about a case where there is one dist-git repo "downstream1", that contains several top-level go repos, "upstream1", "upstream2", and "upstream3".

This may just be terminology nitpicking, but it is untrue that the UPSTREAM relationship is limited to one. A SL can only have one component that it is the SOURCE_FOR, but it can have multiple SL UPSTREAMs.

What I would expect from my understanding of the schema is:
1) There is a SL for the dist-git repo and a component for "downstream1", and they are linked via the SOURCE_FOR relationship.
2) There are 3 SLs and Components for "upstream1..3", each linked appropriately with a SOURCE_FOR relationship.
3) The "downstream1" SL and the "upstream1..3" SLs are linked via the UPSTREAM relationship.
4) The venderoed sources in each of the top-level go repos each are given a Component and a SL, and then that SL is linked to the "upstreamx" SL via the EMBEDS relationship.

Does that sound right you to @mprahl @twaugh?

Reply:
It does -- but are we all happy that the SUPERSEDES relationships will work out correctly in that scenario?

Reply:
It should. The SUPERSEDES relationship is calculated based on "other SLs that relate to this Component". So the different vendored SLs will never supersede each other, only other versions of that same component. If you just use create_or_update_source_location it should take care of that for you, as well as connecting the SOURCE_FOR relationship with the component you pass in.

Reply:
Updated it to work this way. Thanks!